url29ecd5d728_geodb_db3dcore_tgz-pJ8-de_uos_igf_db3d_dbms_model3d_SegmentNetBuilderJ8/src/de/uos/igf/db3d/dbms/geom/Tetrahedron3D.java

```diff
@@ -0,0 +1,2623 @@
+		// triangle) with triangle
+		if (!(this.getMBB().intersects(triangle.getMBB(), sop)))
+			return null;
+
+		SimpleGeoObj planeIntersectionResult = this.intersection(triangle
+				.getPlane(sop), sop);
+		if (planeIntersectionResult != null) {
+			switch (planeIntersectionResult.getType()) {
+
+			case SimpleGeoObj.POINT3D:
+				Point3D p = (Point3D) planeIntersectionResult;
+				if (triangle.containsInPlane(p, sop))
+					return p;
+				else
+					return null;
+
+			case SimpleGeoObj.SEGMENT3D:
+				return triangle.intersectionInPlane(
+						((Segment3D) planeIntersectionResult), sop);
+
+			case SimpleGeoObj.TRIANGLE3D:
+				return triangle.intersectionInPlane(
+						((Triangle3D) planeIntersectionResult), sop);
+
+			case SimpleGeoObj.WIREFRAME3D:
+				Wireframe3D wf = (Wireframe3D) planeIntersectionResult;
+				return intersectionInPlane(triangle, wf, sop);
+
+			default: // no intersection
+				return null;
+			}
+		} else
+			return null;
+	}
+
+	/*
+	 * Special method to compute the intersection of a given triangle and the
+	 * given wireframe which has to be triangulatable into two triangles (flat
+	 * wireframe with four segments) in the same plane as the given triangle.
+	 * Method is used in .projection(triangle) and .intersection(triangle).
+	 * 
+	 * Returns <code>null</code> if no intersection occures. Returns Point3D,
+	 * Segment3D, Triangle3D or Wireframe3D dependent on how the triangle and
+	 * the wireframe intersect.
+	 * 
+	 * @param triangle Triangle3D
+	 * 
+	 * @param wf Wireframe3D (must be triangulatable into two triangles lying in
+	 * the same plane as triangle)
+	 * 
+	 * @return SimpleGeoObj - result of intersection.
+	 * 
+	 * @throws IllegalStateException - if the intersectsInt(Line3D line,
+	 * ScalarOperator sop) method of the class Line3D (which computes the
+	 * intersection of two lines) called by this method returns a value that is
+	 * not -2, -1, 0 or 1.
+	 * 
+	 * @throws IllegalStateException - signals Problems with the dimension of
+	 * the wireframe.
+	 * 
+	 * @throws IllegalStateException - if the result of the intersection is not
+	 * a 3D simplex.
+	 * 
+	 * @throws IllegalArgumentException - if validation of a Triangle3D fails.
+	 * The exception originates in the constructor Triangle3D(Point3D, Point3D,
+	 * Point3D, ScalarOperator).
+	 * 
+	 * @throws IllegalArgumentException - if validation of a Triangle3D fails.
+	 * The exception originates in the constructor Triangle3D(Point3D, Point3D,
+	 * Point3D, ScalarOperator).
+	 * 
+	 * @throws IllegalArgumentException - if index of a triangle point is not 0,
+	 * 1 or 2. The exception originates in the method getPoint(int) of the class
+	 * Triangle3D.
+	 * 
+	 * @throws ArithmeticException - if norm equals zero in epsilon range. This
+	 * exception originates in the method normalize(ScalarOperator) of the class
+	 * Vector3D.
+	 */
+	private SimpleGeoObj intersectionInPlane(Triangle3D triangle,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 1962-1974, lines 2018-2035 into separate methods.